cpp/src/binary/jit/util/type.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2018-2019 BlazingDB, Inc.
+ *     Copyright 2018 Christian Noboa Mardini <christian@blazingdb.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "binary/jit/util/type.h"
+
+namespace gdf {
+namespace binops {
+namespace jit {
+
+    const char* getTypeName(gdf_dtype type) {
```

Comment:
This looks like code that needs to be replaced with the `type_dispatcher`.

Reply:
I thought of doing that but considering the amount of code there is in each `case`, it'd be much uglier and unnecessarily complex if made with multiple template specialized functions. For context, the only use case is [Launcher::instantiate](https://github.com/devavret/cudf/blob/4413a78841b9990e4bfc3a6907a1112329ebd783/cpp/src/binary/jit/core/launcher.h#L55)

Reply:
The problem is that if it doesn't use the `type_dispatcher`, then now this switch case has to be maintained such that if a new `gdf_dtype` is added, someone needs to remember to go update this code.

Reply:
Which we would need to do regardless. Because this code lives here and someone would need to remember to add template specializations for each new type added. Unless you mean I should **move** this to `type_dispatcher`...

Reply:
(Per slack convo)

This should be updated to use the type_dispatcher with a function that stringifys a type as shown here: https://stackoverflow.com/questions/81870/is-it-possible-to-print-a-variables-type-in-standard-c/20170989#20170989

The C++14 solution can be used as-is or C++17 solution could work if modified to just return a `const char*`. Also note that there is excessive portability macros that we can remove.

Reply:
I tried using the solution given in this SO answer and ran into some issues:

This relies on `__PRETTY_FUNCTION__`, (which is a compiler extension and not part of standard). And so, the string generated depended on the location and signature of the code. e.g. 
```c++
std::string cudf::binops::jit::type_name::operator()() [with T = float; std::string = std::basic_string<char>]
```
And, the extraction of type names uses magic numbers.
```c++
return std::string(p.data() + 65, p.size() - 105);
```
If/when we refactor this, those magic numbers would need to be manually tuned, as I found out myself. Worse, it will silently build and only break when the end user tries to run a JIT operation.
I'm struggling to find a more proper solution and this one seems bad for the above reasons.

Apart from this, I also ran into issues with the custom implementation of `common_type` which couldn't understand the type names generated from this function but those will hopefully go away when I replace it with the standard implementation in Jitify.

Reply:
> This relies on `__PRETTY_FUNCTION__`, (which is a compiler extension and not part of standard). And so, the string generated depended on the location and signature of the code. e.g.

Indeed, that is a bit annoying, but assuming that it lives in one place, and we tune those magic numbers accordingly, it _should_ be stable. 

> If/when we refactor this, those magic numbers would need to be manually tuned, as I found out myself. Worse, it will silently build and only break when the end user tries to run a JIT operation.

I think the solution here is just proper unit testing.

Curious what @eyalroz thinks.

Reply:
One solution to avoid the magic numbers is to search the string returned from `__PRETTY_FUNCTION__` for the pattern `T=` and then extract the string between `T=` and the next whitespace character.

Reply:
So,

1. CUDA doesn't even work other than with gcc, MSVC or, differently, with clang. So we're not effectively targeting other compilers anyway, for now.
2. Instead of `__PRETTY_FUNCTION__`, which works at compile-time, we can use the C++11 version and the `<cxxabi.h>` header. That too is not supported by all compilers, but at least there's no use of magic numbers there.
3. The magic numbers are not than much worse than the assumption that we should look for the `T=`; but it would look nicer I guess. Hinnant links to [such an answer](https://stackoverflow.com/a/35943472/1593077)